File: 2019/OWB/Day6.py
def readInput():
    filepath = 'Day6.txt'
    edges = []
    with open(filepath) as fp:
        line = fp.readline()
        while (line):
            n = line.strip().split(')')
            edges.append((n[0], n[1]))
            line = fp.readline()
    return edges
    
def buildGraph(edges):
    graph = {}
    startNode = next((x for x in edges if x[0] == 'COM'), None)
    addEdgeToGraph(startNode, 0, edges, graph)
    return graph

def addEdgeToGraph(edge, distance, edges, graph):
    if edge[0] not in graph:
        graph[edge[0]] = []
    graph[edge[0]].append((edge[1], distance+1))
    
    for e in [x for x in edges if x[0] == edge[1]]:
        addEdgeToGraph(e, distance + 1, edges, graph)
        
def search(edgeFrom, edgeTo, graph):
    searchList = [edgeFrom]
    visited = [edgeFrom]
    while len(searchList) > 0:
        
    

def puzzle1():
    edges = readInput()
    graph = buildGraph(edges)
    sum = 0
    for k in graph:
        for edge in graph[k]:
            sum += edge[1]
    return sum

def puzzle2():
    return

######### PUZZLE 6 #########
print("Day 6:", puzzle1(), puzzle2())

File: 2019/ROX/advent/src/dk/alfabetacain/advent_of_code/day6/Day6.scala
package dk.alfabetacain.advent_of_code.day6

object Day6 {

  def puzzle1(): Unit = {
    val lines = io.Source
      .fromFile("advent/resources/day6/puzzle_1.txt")
      .getLines
      .filter(!_.isEmpty)
      .toList

    val graph = buildGraph(lines)

    val withDistances = bfs(graph.get("COM").get, graph)

    println("puzzle1 sum = " + withDistances.values.sum)
  }

  def puzzle2(): Unit = {
    val lines = io.Source
      .fromFile("advent/resources/day6/puzzle_1.txt")
      .getLines
      .filter(!_.isEmpty)
      .toList

    val graph = buildGraph(lines)

    val withDistances = bfs(graph.get("YOU").get, graph)
    // -2 because we move from our orbit to SAN's
    println(
      "puzzle 2 distance: " + (withDistances.get(graph.get("SAN").get).get - 2)
    )
  }

  final case class Vertex(name: String, neighbours: Set[String])

  def buildGraph(lines: List[String]): Map[String, Vertex] = {
    val map = Map.empty
    lines.foldLeft[Map[String, Vertex]](Map.empty) { (map, line) =>
      val parts = line.split(')')
      val first: Vertex =
        map.get(parts(0)).getOrElse(Vertex(parts(0), Set.empty))
      val second: Vertex =
        map.get(parts(1)).getOrElse(Vertex(parts(1), Set.empty))
      (map + (parts(0) -> first.copy(neighbours = first.neighbours + second.name
      ))) + (parts(1) -> second.copy(neighbours = second.neighbours + first.name
      ))
    }
  }

  def bfs(from: Vertex, all: Map[String, Vertex]): Map[Vertex, Int] = {
    var i = 0
    var toBeExplored = Set(from)
    var explored = Map.empty[Vertex, Int]
    while (!toBeExplored.isEmpty) {
      val res = toBeExplored
        .foldLeft[(Map[Vertex, Int], Set[Vertex])]((explored, Set.empty)) {
          case ((m, next), vertex) =>
            m.get(vertex) match {
              case None =>
                (
                  m + (vertex -> i),
                  next ++ vertex.neighbours.map(n => all.get(n).get)
                )
              case Some(_) =>
                (m, next)
            }
        }
      explored = res._1
      toBeExplored = res._2
      i += 1
    }
    explored
  }
}
